File: Data Structures & Algorithms/combination-target-sum/submission-5.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        sol, res = [], []

        def backtrack(i, summ):
            if summ == target:
                res.append(sol[:])
                return 

            for j in range(i, len(nums)):
                if summ + nums[j] > target:
                    continue
                sol.append(nums[j])
                backtrack(j, nums[j] + summ)
                sol.pop()


        backtrack(0, 0)
        return res

File: Data Structures & Algorithms/permutations-ii/submission-5.go
func permuteUnique(nums []int) [][]int {
	count := make(map[int]int)
	for _, n := range nums {
		count[n]++
	}

	sol := []int {} 
	res := [][]int {}
	var backtrack func() 

	backtrack = func() {
		if len(sol) == len(nums) {
			res = append(res, append([]int{}, sol...))
		}

		for val, c := range count {
            if c == 0 {
				continue
			}

			count[val]--
			sol = append(sol, val)
			backtrack()

			sol = sol[:len(sol) - 1]
			count[val]++
		}
	}

	backtrack()
	return res
}
